File: vpon_plugin_poc/android/src/main/kotlin/io/flutter/plugins/vponmobileads/VponFlutterAdLoadedListener.java
package io.flutter.plugins.vponmobileads;

import com.vpon.ads.VponAdListener;
import com.vpon.ads.VponNativeAd;

import java.lang.ref.WeakReference;

interface VponFlutterAdLoadedListener {
    void onAdLoaded();
}

class VponFlutterAdListener extends VponAdListener {

    protected final int adId;
    protected final VponAdInstanceManager adInstanceManager;

    VponFlutterAdListener(int adId, VponAdInstanceManager adInstanceManager) {
        this.adId = adId;
        this.adInstanceManager = adInstanceManager;
    }

}

class VponFlutterBannerAdListener extends VponFlutterAdListener {

    final WeakReference<VponFlutterAdLoadedListener> adLoadedListenerWeakReference;

    VponFlutterBannerAdListener(int adId, VponAdInstanceManager adInstanceManager
            , VponFlutterAdLoadedListener vponFlutterAdLoadedListener) {
        super(adId, adInstanceManager);
        adLoadedListenerWeakReference = new WeakReference<>(vponFlutterAdLoadedListener);
    }

    @Override
    public void onAdLoaded() {
        adLoadedListenerWeakReference.get().onAdLoaded();
    }
}

class VponFlutterNativeAdListener extends VponFlutterAdListener {

    VponFlutterNativeAdListener(int adId, VponAdInstanceManager adInstanceManager) {
        super(adId, adInstanceManager);
    }
}

class VponFlutterNativeAdLoadedListener extends VponNativeAd.OnNativeAdLoadedListener {

    private final WeakReference<VponFlutterNativeAd>nativeAdWeakReference;
    VponFlutterNativeAdLoadedListener(VponFlutterNativeAd flutterNativeAd) {
        nativeAdWeakReference = new WeakReference<>(flutterNativeAd);
    }

    @Override
    public void onNativeAdLoaded( VponNativeAd nativeAd) {
        if (nativeAdWeakReference.get() != null) {
            nativeAdWeakReference.get().onNativeAdLoaded(nativeAd);
        }
    }

}

File: vpon_plugin_poc/android/src/main/kotlin/io/flutter/plugins/vponmobileads/VponFlutterAdLoadedListener.kt
package io.flutter.plugins.vponmobileads

import com.vpon.ads.VponAdListener
import com.vpon.ads.VponAdRequest
import com.vpon.ads.VponNativeAd
import java.lang.ref.WeakReference

interface VponFlutterAdLoadedListener {
    fun onAdLoaded()
}

internal open class VponFlutterAdListener(
    protected val adId: Int,
    protected val vponAdInstanceManager: VponAdInstanceManager
) : VponAdListener()

internal class VponFlutterBannerAdListener(
    adId: Int,
    vponAdInstanceManager: VponAdInstanceManager,
    vponFlutterAdLoadedListener: VponFlutterAdLoadedListener
) : VponFlutterAdListener(adId, vponAdInstanceManager) {
    private val adLoadedListenerWeakReference = WeakReference(vponFlutterAdLoadedListener)

    override fun onAdLoaded() {
        adLoadedListenerWeakReference.get()?.onAdLoaded()
    }
    override fun onAdFailedToLoad(errorCode: Int) {
        var vponErrorCode : VponAdRequest.VponErrorCode = VponAdRequest.VponErrorCode.NO_FILL
        when (errorCode) {
            0 -> vponErrorCode = VponAdRequest.VponErrorCode.INTERNAL_ERROR
            1 -> vponErrorCode = VponAdRequest.VponErrorCode.INVALID_REQUEST
            2 -> vponErrorCode = VponAdRequest.VponErrorCode.NETWORK_ERROR
            3 -> vponErrorCode = VponAdRequest.VponErrorCode.NO_FILL
        }
        vponAdInstanceManager.onAdFailedToLoad(adId,vponErrorCode)
    }
}

internal class VponFlutterNativeAdListener(
    adId: Int,
    vponAdInstanceManager: VponAdInstanceManager
) : VponFlutterAdListener(adId, vponAdInstanceManager)

internal class VponFlutterNativeAdLoadedListener(flutterNativeAd: VponFlutterNativeAd) :
    VponNativeAd.OnNativeAdLoadedListener() {
    private val nativeAdWeakReference = WeakReference(flutterNativeAd)

    override fun onNativeAdLoaded(nativeAd: VponNativeAd?) {
        nativeAdWeakReference.get()?.onNativeAdLoaded(nativeAd)
    }
}